### ISSUE TITLE
Browser security issue prevents workbench from displaying

### ISSUE THREAD

**Issue by jaamp** [SharePoint/sp-dev-docs#4508]
#### Category
- [x] Bug
#### Expected or Desired Behavior
The local workbench should open without any error when running the gulp serve.
#### Observed Behavior
Will not connect to HTTPS://localhost:xxxx/temp/workbench.html due to “the site uses outdated or unsafe TLS security settings”.  This occurs using Edge, Chrome, Firefox
#### Steps to Reproduce
Install SPFx framework following the instructions on https://docs.microsoft.com/en-us/sharepoint/dev/spfx/set-up-your-development-environment
Using windows 10 home. (although I am getting the same error on my work computer which is Win 10 professional, and I have admin privileges on that computer as well – and ran everything on that computer in admin mode.
>
#### What I have done…
1.	Followed instructions on MS page above – did not work
2.	Ran in Admin mode  - did not work
3.	Added certs to trusted root – did not work
4.	Set all browsers to accept all forms of TLS – did not work
5.	Basically, I have tried everything I have found on web searches.. save reinstalling windows.

Thanks for your contribution! Sharing is caring.

**Comment by BassemNKhalil** [comment]
Are you able to share a bit more information on the environment, i.e your SharePoint framework version as well as the targeted SharePoint version (SP16, SP19 or SPO)?

**Comment by jaamp** [comment]
My Targeted SP Version is SP16, not sure the framework version, however, the issue is not with the targeted version of SP or the framework.  I can't even get to the workbench.  I get a certificate error / site not trusted error.

**Comment by BassemNKhalil** [comment]
Please check #4485 - it seems to be the same scenario. I’ve managed to workaround it by changing the configuration in the serve.json config file.

**Comment by PopWarner** [comment]
You can always check your installed version of SPFx by running the following command in your console.
`npm list --depth=0 -g`

You'll be looking for the line that says
`@microsoft/generator-sharepoint@X.X.X`

:)

**Comment by jaamp** [comment]
oh, okay...  maybe this helps..

NPM version - 6.9.0
Node version - 10.16.3
Yoeman version - 3.1.0
Glup CLI version - 2.2.0
Microsoft-Generator version - 1.9.1

**Comment by jaamp** [comment]
here is some more information

PS C:\Users\jotow\Desktop\sp> gulp trust-dev-cert
Build target: DEBUG
[07:53:24] Using gulpfile ~\Desktop\sp\gulpfile.js
[07:53:24] Starting gulp
[07:53:24] Starting 'trust-dev-cert'...
[07:53:24] Starting subtask 'trust-cert'...
[07:53:24] Finished subtask 'trust-cert' after 77 ms
[07:53:24] Finished 'trust-dev-cert' after 78 ms
[07:53:25] ==================[ Finished ]==================
[07:53:25] Project sp version: 0.0.1
[07:53:25] Build tools version: 2.5.3
[07:53:25] Node version: v10.16.3
[07:53:25] Total duration: 3.17 s
PS C:\Users\jotow\Desktop\sp> gulp serve
Build target: DEBUG
[07:53:48] Using gulpfile ~\Desktop\sp\gulpfile.js
[07:53:48] Starting gulp
[07:53:48] Starting 'serve'...
[07:53:48] Starting subtask 'pre-copy'...
[07:53:48] Finished subtask 'pre-copy' after 11 ms
[07:53:48] Starting subtask 'copy-static-assets'...
[07:53:48] Starting subtask 'sass'...
[07:53:49] Finished subtask 'copy-static-assets' after 766 ms
[07:53:49] Finished subtask 'sass' after 757 ms
[07:53:49] Starting subtask 'tslint'...
[07:53:49] Starting subtask 'typescript'...
[07:53:49] [typescript] TypeScript version: 2.2.2
[07:53:50] Finished subtask 'tslint' after 1.25 s
[07:53:50] Finished subtask 'typescript' after 1.25 s
[07:53:50] Starting subtask 'ts-npm-lint'...
[07:53:50] Finished subtask 'ts-npm-lint' after 18 ms
[07:53:50] Starting subtask 'api-extractor'...
[07:53:50] Finished subtask 'api-extractor' after 946 μs
[07:53:50] Starting subtask 'post-copy'...
[07:53:50] Finished subtask 'post-copy' after 706 μs
[07:53:50] Starting subtask 'collectLocalizedResources'...
[07:53:50] Finished subtask 'collectLocalizedResources' after 3.99 ms
[07:53:50] Starting subtask 'configure-webpack'...
[07:53:51] Finished subtask 'configure-webpack' after 591 ms
[07:53:51] Starting subtask 'webpack'...
(node:13396) DeprecationWarning: loaderUtils.parseQuery() received a non-string value which can be problematic, see https://github.com/webpack/loader-utils/issues/56
parseQuery() will be replaced with getOptions() in the next major version of loader-utils.
[07:53:51] Finished subtask 'webpack' after 570 ms
[07:53:51] Starting subtask 'configure-webpack-external-bundling'...
[07:53:51] Finished subtask 'configure-webpack-external-bundling' after 899 μs
[07:53:51] Starting subtask 'copy-assets'...
[07:53:51] Finished subtask 'copy-assets' after 47 ms
[07:53:51] Starting subtask 'write-manifests'...
[07:53:52] Finished subtask 'write-manifests' after 510 ms
[07:53:52] Starting subtask 'serve'...
Starting api server on port 5432.
Registring api: /getwebparts
Registring api: /*.*
Registring api: /workbench
[07:53:52] Finished subtask 'serve' after 434 ms
[07:53:52] Finished 'serve' after 4.24 s
[07:53:52] Server started https://localhost:4321
[07:53:52] LiveReload started on port 35729
[07:53:52] Opening https://localhost:5432/workbench using the default OS app
[07:53:53] ==================[ Finished ]==================
  Request: [::1] '/workbench'
[07:53:53] Project sp version: 0.0.1
[07:53:53] Build tools version: 2.5.3
[07:53:53] Node version: v10.16.3
[07:53:53] Total duration: 7.15 s

**Comment by jaamp** [comment]
@BassemNKhalil  -  I went in and edited the serve.json to below...
{
  "$schema": "https://developer.microsoft.com/json-schemas/core-build/serve.schema.json",
  "port": 4321,
  "https": false,
  "initialPage": "http://localhost:5432/workbench",
  "api": {
    "port": 5432,
    "entryPath": "node_modules/@microsoft/sp-webpart-workbench/lib/api/"
  }
}

**Comment by jaamp** [comment]
I was able to get to the workbench using just http... so, my next question, is it a big deal that I cannot get to the https?  Will I still be able to fully develop webparts without the https?

**Comment by BassemNKhalil** [comment]
Hi @jaamp, the settings in serve.json are only used when you’re using gulp serve so it doesn’t affect the WebPart when deployed later on to SharePoint.

Since I found #4485 I’ve been developing a WebPart locally on the workbench then deployed it for good measures to SharePoint online, SP2016 as well as Microsoft teams. 

I’m hoping someone from the community can shed some light as to why this behaviour happens and hopefully Microsoft can change it in a following release.

**Comment by andrewconnell** [comment]
@jaamp 
Have you taken a look at the first two entries under the **Troubleshooting** section here?
https://docs.microsoft.com/en-us/sharepoint/dev/spfx/set-up-your-development-environment#troubleshooting

When you are creating the project, even though it's running on your local workbench, did you choose the SP2016 environment when creating the project? If so, check that first one.

Changing the local webserver to use HTTP and not HTTPS, while it may get around the issue, doesn't fix it. It is not really a solution IMHO...

**Comment by jaamp** [comment]
Hi, 

Prior to starting I removed the .gcb-serve-data folder. Set the Environment Variable to: NODE_NO_HTTP2=1

Then when creating the project is selected, in Yeoman, "SharePoint 2016 onwards, including 2019 and SharePoint Online"

ran glup trust-dev-cert

than glup serve -- still no joy

**Comment by andrewconnell** [comment]
By chance have you tried a SharePoint Online target environment and then tried running the local workbench?

**Comment by andrewconnell** [comment]
Couple other ideas... chance you are behind a proxy?

Did you have your A/V enabled when you installed the original NPM dependencies (SPFx generator, Yeoman, Gulp) or when you initially created the project and ran `npm install`?

**Comment by jaamp** [comment]
Apologize for the delayed response, was finalizing other projects... So no proxy, did a toolchain build with A/V and Firewall off .. still did not work in local host.  From the same build, I changes the gulp serve config to point to my SP dev tenant workbench, this seems to work with the https://<mytenet>.sharepoint.com/sites/<developersite>/_layouts/15/workbench.aspx

So maybe that might be the only way I will get there is to work directly in the SP instead of localhost.

**Comment by andrewconnell** [comment]
Have you been able to navigate to localhost for other projects (like ASP.NET)?

**Comment by jaamp** [comment]
Hi, I am unable to get to HTTPS with ASP.NET (same error site cannot provide secured connection).  I also went to extremes... I completely reset my windows 10.  Then downloaded only VS code, Node LTS, and CMDER, did the tool chain, and still received the same results...

I am sure I am missing something...

**Comment by andrewconnell** [comment]
In that case this isn't specific to SPFx, rather it's something going on with your machine. When you say ASP.NET, unclear if you are talking about doing this from VSCode (dotnet core) or Visual Studio.  Regardless, I saw this that has a lot of steps on how to investigate and trust a cert from the browser.

See if this helps... we have a lot of these issues in the list & I'm trying to find common root causes to create the definitive guide on how to resolve it... please let me know how this works for you. You can likely ignore the IIS Express stuff, rather use the parts that have you inspect the certs installed, deleting them, and manually trusting it from the browser:

https://www.pluralsight.com/guides/visual-studio-2017-resolving-ssl-tls-connections-problems-with-iis-express

**Comment by andrewconnell** [comment]
Those are two completely different certs. Visual Studio 2019 has it's own... there isn't a VS Code cert... an NPM package that the SPFx project is relying on via the workbench has it's own cert. 

My point was to follow the same process of manually trusting the cert via VS Code & the browser that the article went through with a Visual Studio solution.

**Comment by jaamp** [comment]
I have found the issue at least with me... there is a bug in the @microsoft-gulp-core-build-serve.  In
\node_modules\@microsoft\gulp-core-build-serve\lib\certificates.d.ts
the first line is
/// <reference path="../src/NodeForgeExtensions.d.ts" /> 
it should be
// <reference path="../src/NodeForgeExtensions.d.ts" />

the three slashed are like a double negative.. where I think MS wanted to comment out, it actually re-commented it in.... anyway, I removed the extra forward slash, saved the file, ran gulp trust-dev-cert, then gulp serve.. and like a charm the workbench opened in my local host https... 

You can consider this closed and I hope this helps others...

**Comment by jaamp** [comment]
Never letting things be... here is an addendum... The above error I identified, is not the problem, it is still an error, but not the problem.  When AC asked earlier "By chance have you tried a SharePoint Online target environment and then tried running the local workbench?" it failed.  However, when testing today that is what worked, building a 365 only with edits to

_\node_modules@microsoft\gulp-core-build-serve\lib\certificates.d.ts_

So, I built a new 365 only without making edits to the .. _\node_modules ... and it still worked,,, great, so the problem was not the error in ... _\node_modules ... .  

Then I build a 2019 and above, again built like a charm https and everything, again no changes to the  ... _\node_modules ... .

Finally, I build a 2016 and above,  there I ran into the cert issue again.  I tried it with changing, and with out changing  ... _\node_modules ... .

I don't think the issue is with my computer or trusting the certs, it is somewhere deep in the @microsoft-gulp-core-build, way beyond my level of experience. 

as another addendum - when I was looking deeper into the @microsoft-gulp-core-build I found a few more errors in the "Rush" area (not sure what that is)

I don't want to re-open the issue, I can work fine with the 365, for my work, we will be on 2016 prem at the end of the year, and by end of next year we plan to be on 365, so what I learn now is good, and will set me up for our eventual move.

**Comment by andrewconnell** [comment]
I think I finally figured this one out: the `NODE_NO_HTTP2` environment variable was added in NodeJS v8, but removed in NodeJS v10. I can't find anything in v10+ that enables a user to force HTTP1 from the CLI. Thus... you are stuck using NodeJS v8 if you are using SPFx v1.1 for SharePoint Server 2016.

I've updated the SPFx docs & they've been merged to master, but won't go live for a few more days.

More: 
https://www.andrewconnell.com/blog/spfx-which-version-of-spfx-generator-to-install/
